### PR TITLE
refactor(core): Remove `isObservable()` in favor `isSubscribable()`.

### PR DESCRIPTION
**Pull request by JeanMeche**
The private util `isObservable` was actually just testing the same thing as`isSubscribable()`. As the implementation is closer to the function's name, let's only keep ``isSubscribable`.

This commit also adds generic typing to improve the return type of the guard.

## PR Type
What kind of change does this PR introduce?

- [x] Refactoring (no functional changes, no api changes)

### CODE REVIEW

**Review by dylhunn — APPROVED** [review]
reviewed-for: fw-core, fw-common

We're losing a bit of language precision here -- since technically Observable has some extra methods -- but there is no behavior change, and we are saving some bundle size, so I think it's worth it.

**Comment by dylhunn** [comment]
Presubmit in http://cl/513595632

**Comment by dylhunn** [comment]
caretaker: presubmit failures are flakes

**Comment by dylhunn** [comment]
Just FYI, @pkozlowski-opensource has asked to review this also!

File: packages/core/src/util/lang.ts [hunk 1]
```diff
@@ -18,20 +18,8 @@ export function isPromise<T = any>(obj: any): obj is Promise<T> {
 }
 
 /**
- * Determine if the argument is a Subscribable
+ * Determine if the argument is an Observable
  */
-export function isSubscribable(obj: any|Subscribable<any>): obj is Subscribable<any> {
+export function isObservable<T>(obj: any|Subscribable<T>): obj is Observable<T> {
   return !!obj && typeof obj.subscribe === 'function';
```
Comment (pkozlowski-opensource) [inline packages/core/src/util/lang.ts:24]: The issue here is that the logic test for the presence of  the `subscribe` method but then assumes (in the type system) that the object is of type `Observable`. This is not true in general as you can have object with the subscribe method that are _not_ observables. This distinction is important and was specifically introduced in #39627 to narrow down type to https://rxjs.dev/api/index/interface/Subscribable

File: packages/common/src/pipes/async_pipe.ts [hunk 2]
```diff
@@ -139,7 +139,7 @@ export class AsyncPipe implements OnDestroy, PipeTransform {
       return _promiseStrategy;
     }
 
-    if (ɵisSubscribable(obj)) {
+    if (ɵisObservable(obj)) {
```
Comment (pkozlowski-opensource) [inline packages/common/src/pipes/async_pipe.ts:142]: We really want to test for subscribable only here.

**Review by pkozlowski-opensource — CHANGES_REQUESTED** [review]
see https://github.com/angular/angular/pull/49295/files#r1124750317

File: packages/core/src/util/lang.ts (hunk 1 above)
Comment (JeanMeche) [reply packages/core/src/util/lang.ts:24]: Fair enough, I've just tested the suggestion of @divdavem  in his PR. We can actually drop `isObservable` and only use `isSubscribable` which matches what the code actually checks !

File: packages/core/src/util/lang.ts [hunk 3]
```diff
@@ -18,20 +18,8 @@ export function isPromise<T = any>(obj: any): obj is Promise<T> {
 }
 
 /**
- * Determine if the argument is a Subscribable
+ * Determine if the argument is an Observable
```
Comment (pkozlowski-opensource) [inline packages/core/src/util/lang.ts:21]: I guess you want to revert this change now?

File: packages/core/src/util/lang.ts [hunk 4]
```diff
@@ -18,20 +18,8 @@ export function isPromise<T = any>(obj: any): obj is Promise<T> {
...
  */
-export function isSubscribable(obj: any|Subscribable<any>): obj is Subscribable<any> {
+export function isSubscribable<T>(obj: any|Subscribable<T>): obj is Observable<T> {
```
Comment (pkozlowski-opensource) [inline packages/core/src/util/lang.ts:23]: `is Subscribable<T>`, right?

File: packages/router/src/utils/collection.ts [hunk 5]
```diff
@@ -60,7 +60,7 @@ export function last<T>(a: T[]): T|null {
 
 
 export function wrapIntoObservable<T>(value: T|Promise<T>|Observable<T>): Observable<T> {
```
Comment (pkozlowski-opensource) [inline packages/router/src/utils/collection.ts:62]: Here the check doesn't match the return value - we are testing for the presence of `subscribe` but assuming the return type to be `Observable`

**Review by pkozlowski-opensource — APPROVED** [review]

**Review by atscott — APPROVED** [review]

**Comment by AndrewKushnir** [comment]
@pkozlowski-opensource @atscott ok to mark this PR as ready for merge?

**Comment by jessicajaniuk** [comment]
We'll have to update to `target: minor` since this has conflicts with `15.2.x`. @JeanMeche Feel free to create a patch PR to target that branch if you'd like to see this land there, too.

**Review by AndrewKushnir — APPROVED** [review]

**Comment by jessicajaniuk** [comment]
This PR was merged into the repository by commit daaf0fd2f6fefc0bf612e47604cb3cdcd3305fed.
